openupgradelib/openupgrade.py

```diff
@@ -960,19 +962,35 @@ def reactivate_workflow_transitions(cr, transition_conditions):
             (condition, transition_id))
 
 
-def migrate(no_version=False):
+def migrate(no_version=False, v9env=None, context=None, *args, **kwargs):
     """
-    This is the decorator for the migrate() function
-    in migration scripts.
-    Set argument 'no_version' to True if the method as to be taken into account
-    if the module is installed during a migration.
-    Return when the 'version' argument is not defined and no_version is False,
-    and log execeptions.
-    Retrieve debug context data from the frame above for
-    logging purposes.
+    This is the decorator for the migrate() function in migration scripts.
+    It returns if `version` is not defined and `no_version` is False,
+    and logs execeptions. It also retrieves debug context data from the
+    frame above for logging purposes. Optionally it provides a fully fledged
+    Odoo Environment (with `v9env` = True). An arbitrary context can be
+    preloaded into the Environment via `context` dict.
+
+    :param no_version: Set to `True` if the migrate method has to be taken
+    into account if the module is installed during a migration.
+    :param v9env: Set to `True` gives you a fully fledged Odoo Environment.
+    As per v10, you automatically get it, but can disable with `False`.
+    The wrapped function signature changes to `migrate(env, *args, **kwargs):`
+        Obtain Version: env.context['migrate_version']
+        Obtain Cursor: env.cr
+        To genuinely support calling legacy methods you should do:
+        ```
+        def migrate(env, *args, **kwargs):
+            version = env.context['migrate_version']
+            cr = env.cr
+        ```
+    :param context: An arbitrary odoo context you might want to set up. You
+    can access this through `env.context`, where the Environment is present.
     """
     def wrap(func):
-        def wrapped_function(cr, version):
+        @wraps(func)
+        def wrapped_function(cr, version, no_version=no_version,
+                             context=context, *args, **kwargs):
```

Comment:
@StefanRijnhart Here was another one `env=env` was missing. It looks like this has become necessary because of the way the `@wraps` decorator works.